File: admin-backend/admin-shared/admin-shared-persistence/src/main/java/io/github/faustofan/admin/shared/persistence/interceptor/TenantDraftInterceptor.java
package io.github.faustofan.admin.shared.persistence.interceptor;

import io.github.faustofan.admin.shared.common.constant.CustomConstants;
import io.github.faustofan.admin.shared.common.context.AppContextHolder;
import io.github.faustofan.admin.shared.common.exception.CommonErrorCode;
import io.github.faustofan.admin.shared.common.exception.SystemException;
import io.github.faustofan.admin.shared.persistence.model.TenantAware;
import io.github.faustofan.admin.shared.persistence.model.TenantAwareDraft;
import io.github.faustofan.admin.shared.persistence.model.TenantAwareProps;
import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Component;

@Component
public class TenantDraftInterceptor implements DraftInterceptor<TenantAware, TenantAwareDraft> {
    @Override
    public void beforeSave(@NotNull TenantAwareDraft draft, @Nullable TenantAware original) {
        if(null == AppContextHolder.getContext())
            throw new SystemException(CommonErrorCode.CONTEXT_MISSING);

        Long tenantId = AppContextHolder.getContext().tenantId();
        long userId = AppContextHolder.getContext().userId() == null
                ? CustomConstants.SYSTEM_USER_ID
                : AppContextHolder.getContext().userId();

        if(null == tenantId)
            throw new SystemException(CommonErrorCode.TENANT_ID_MISSING);

        // 新建时，设置租户ID
        if(null == original) {
            if(!ImmutableObjects.isLoaded(draft, TenantAwareProps.TENANT_ID)) {
                draft.setTenantId(tenantId);
            }
            if(!ImmutableObjects.isLoaded(draft, TenantAwareProps.CREATED_BY)) {
                draft.setCreatedBy(userId);
            }
        }
        // 更新时，都设置更新人和更新时间
        draft.setUpdatedBy(userId);
    }
}

File: src/main/kotlin/io/github/faustofan/admin/config/jimmer/TenantDraftInterceptor.kt
package io.github.faustofan.admin.config.jimmer

import io.github.faustofan.admin.common.context.AppContextHolder
import io.github.faustofan.admin.common.model.TenantAware
import io.github.faustofan.admin.common.model.TenantAwareDraft
import org.babyfish.jimmer.kt.isLoaded
import org.babyfish.jimmer.sql.DraftInterceptor
import org.springframework.stereotype.Component

/**
 * 租户字段拦截器
 *
 * 在实体保存前自动填充租户ID
 * 从当前用户上下文中获取租户信息
 */
@Component
class TenantDraftInterceptor : DraftInterceptor<TenantAware, TenantAwareDraft> {

    override fun beforeSave(draft: TenantAwareDraft, original: TenantAware?) {
        // 只在新增时填充租户ID
        if (original === null) {
            if (!isLoaded(draft, TenantAware::tenantId)) {
                val tenantId = AppContextHolder.get().tenantId
                    ?: throw IllegalStateException("租户上下文未初始化，无法保存租户隔离数据")
                draft.tenantId = tenantId
            }
        }
        // 注意：更新时不允许修改租户ID，保证数据安全
    }
}

